lib/player.rb

```diff
@@ -12,44 +12,116 @@ def add_ship(ship)
     @ships << ship
   end
 
+
+
   def ship_setup
     if @type == :auto
-      # helper-method: auto_generation CALLS
-      # helper-method: player_place_ships
-      # auto generation
-      p "type auto"
-    elsif @type == :user
-      # Put in a guard for user error
-
-      # helper-method: player_place_ships
-      @ships.each do |ship|
-        if ship.length == 3
-          puts "Enter the squares for the Cruiser (3 spaces):"
-          puts "> "
-        elsif ship.length == 2
-          puts "Enter the squares for the Submarine (2 spaces):"
+      until all_ships_placed?
+        auto_generate_coordinates
+      end
+    else
+      until all_ships_placed?
+        @ships.each do |ship|
+          puts "Enter the squares for the #{ship.name} (#{ship.length} spaces)"
           puts "> "
-        end
-
-        user_ship_coords = gets.chomp.split(/ /)
-        # Consider loop for user error
-        if @board.valid_placement?(ship, user_ship_coords) == true
-          @board.place(ship, user_ship_coords)
-        elsif @board.valid_placement?(ship, user_ship_coords) == false
-          loop do
-            puts "Those are invalid coordinates. Please try again:"
-            puts "> "
-            user_ship_coords = gets.chomp.split(/ /)
-
-            if @board.valid_placement?(ship, user_ship_coords) == true
-              @board.place(ship, user_ship_coords)
-              require "pry"; binding.pry
-              exit # Could be a helper-method
+          user_ship_coords = gets.chomp.upcase.split(/ /)
+          result = @board.valid_placement?(ship, user_ship_coords)
+          if result == false
+            until this_ship_placed?(ship)
+              puts "Those are invalid coordinates. Please try again: "
+              user_ship_coords2 = gets.chomp.upcase.split(/ /)
+            @board.valid_placement?(ship, user_ship_coords2)
+            @board.validated_placement
             end
+          else
+            @board.validated_placement
           end
         end
       end
     end
   end
 
+  def auto_generate_coordinates
+    @ships.each do |ship|
+      until this_ship_placed?(ship)
```

Comment:
Loving the until loops!!!! Gonna go check out the helper method associated with this.